util-images/network/netperfbenchmark/pkg/worker/worker.go

```diff
@@ -0,0 +1,277 @@
+}
+
+// StartTCPClient starts iperf client for tcp measurements
+func (w *Worker) StartTCPClient(rw http.ResponseWriter, request *http.Request) {
+	klog.Info("Starting TCP Client")
+	w.startWork(rw, request, ProtocolTCP, "iperf", tcpClientArguments)
+}
+
+// StartUDPClient starts iperf client for udp measurements
+func (w *Worker) StartUDPClient(rw http.ResponseWriter, request *http.Request) {
+	klog.Info("Starting UDP Client")
+	w.startWork(rw, request, ProtocolUDP, "iperf", udpClientArguments)
+}
+
+// Handler handles http requests for http measurements
+func (w *Worker) Handler(rw http.ResponseWriter, request *http.Request) {
+	w.sendResponse("hi", rw, http.StatusOK)
+}
+
+func (w *Worker) startWork(rw http.ResponseWriter, request *http.Request, protocol string, command string, arguments []string) {
+	arguments, err := populateTemplates(arguments, request)
+	if err != nil {
+		w.sendResponse(status{Message: err.Error()}, rw, http.StatusBadRequest)
+		return
+	}
+	valMap, err := GetValuesFromURL(request, []string{"timestamp", "duration"})
+	if err != nil {
+		w.sendResponse(status{Message: err.Error()}, rw, http.StatusBadRequest)
+		return
+	}
+	tsint, err := strconv.ParseInt(valMap["timestamp"], 10, 64)
+	if err != nil {
+		klog.Infof("Invalid timestamp: %v %v", valMap["timestamp"], err)
+	}
+	go w.schedule(protocol, tsint, valMap["duration"], command, arguments)
+	w.sendResponse(nil, rw, http.StatusOK)
+	return
+}
+
+func (w *Worker) schedule(protocol string, futureTimestamp int64, duration string, command string, arguments []string) {
+	//If future time is in past,run immediately
+	klog.Infof("About to wait for futuretime: %v", futureTimestamp)
+	klog.Infof("Current time: %v", time.Now().Unix())
+	time.Sleep(time.Duration(futureTimestamp-time.Now().Unix()) * time.Second)
+	w.startedAt = time.Now()
+	w.futureTime = time.Unix(futureTimestamp, 0)
+	w.testDuration = duration
+	w.protocol = protocol
+	w.executeCommand(duration, command, arguments)
+}
+
+func (w *Worker) executeCommand(duration string, commandString string, arguments []string) {
+	command := exec.Command(commandString, arguments...)
+	out, err := command.StdoutPipe()
+	if err != nil {
+		klog.Errorf("unable to obtain Stdout: %v", err)
+		w.err = err
+		close(w.stopCh)
+		return
+	}
+	errorOut, err := command.StderrPipe()
+	if err != nil {
+		klog.Errorf("unable to obtain Stderr: %v", err)
+		w.err = err
+		close(w.stopCh)
+		return
+	}
+	multiOut := io.MultiReader(out, errorOut)
+	err = command.Start()
+	if err != nil {
+		klog.Errorf("command execution failed: %v", err)
+		w.err = err
+		close(w.stopCh)
+		return
+	}
+	w.scanOutput(multiOut)
+}
+
+func (w *Worker) scanOutput(out io.Reader) {
+	scanner := bufio.NewScanner(out)
+	for scanner.Scan() {
+		line := scanner.Text()
+		if len(line) > 0 {
+			w.result = append(w.result, line)
+		}
+	}
+	klog.Info("Command executed,sending result back")
+	if err := scanner.Err(); err != nil {
+		klog.Errorf("Error: %v", err)
+		w.err = err
```

Comment:
Actually, I don't like this pattern - if the error happens later, you may have partial w.result

What you should do is to:
- return ([]string, error) from this function (instead of assigning w.result, w.err)
- in executeCommand do:
```
if result, err := scanOutput(); err == nil {
  w.result = result
} else {
  w.err = err
}
```

Reply:
Made changes to schedule such that it executes command and then parses result and any err during these would be captured , Metrics would inform controller on seeing this error.